### DISCUSSION TITLE
Deeplink to jellyfin's androidtv app

### DISCUSSION THREAD

**Discussion by pheffner** [jellyfin/jellyfin-androidtv#3452] [General]
I'm using the Android TV Remote extension for Home Assistant to build a nice UI for controlling my Sony TVs with ATV. In order to launch apps on the TV, the platform uses deeplinks to start specific ones. An explanation of the deeplink scheme is found at:
[The HA ATV Deeplink Guide](https://community.home-assistant.io/t/android-tv-remote-app-links-deep-linking-guide/567921)
In that doc there is a list of apps which can be launched with their respective link strings. Most allow a URL format or sometimes a string of the form <name>:// e.g. hulu:// or plex://

I tried using the HA developer tool to call the ATV service with parameter strings like jellyfin://  or JellyfinApplication:// which fails. (Using the same call method does launch YouTube, etc. successfuly.)

I retrieved the jellyfin-androidtv sources and looked for the android:host string in all of the AndroidManifest.xml files and found the section which would usually implement the deeplink is configured like this:

       <intent>
            <action android:name="android.intent.action.VIEW" />
            <data
                android:host="youtube.com"
                android:scheme="https" />
        </intent>

...which uses a URL to YouTube. If this were changed to something like:
 
           <intent-filter>
                <action android:name="android.intent.action.VIEW"/>
                <category android:name="android.intent.category.DEFAULT"/>
                <data android:scheme="jellyfin"/>
            </intent-filter>

and pushed out in the next update it would probably implement a proper deeplink to the Jellyfin app. (I believe....)

**Comment by nielsvanvelzen** [comment]
The intent you found is part of a query definition to let the system know we're interested in finding out which apps can play YouTube videos, this is used for trailer playback.

At this time we do not have deeplink support in our Android apps. We do implement the Android search action and the view action (data = item id).

**Reply by pheffner** [reply]
Okay, thanks for your fast response.

**Comment by pandaboy6621** [comment]
What activity can I use to launch jellyfin's android tv version from Home Assistant?

**Reply by ebinf** [reply]
It depends on what you want to achieve. If you **just want to open the app** on your android tv, you could use the following adb command:

```
monkey -p org.jellyfin.androidtv 1
```

The resulting Home Assistant step would be:
```
service: androidtv.adb_command
metadata: {}
data:
  command: monkey -p org.jellyfin.androidtv 1
target:
  entity_id: <YOUR ANDROID TV>
```


But if you, like me, would like to **open a specific title in the app**, you could use the following adb command:
```
 adb shell am start -c android.intent.category.LEANBACK_LAUNCHER -a android.intent.action.VIEW -d <ID OF THE TITLE YOU WANT TO OPEN> -e source 30 org.jellyfin.androidtv/.ui.startup.StartupActivity
```

Here the resulting Home Assistant step would be:
```
service: androidtv.adb_command
data:
  command: >-
    adb shell am start -c android.intent.category.LEANBACK_LAUNCHER -a
    android.intent.action.VIEW -d <ID OF THE TITLE YOU WANT TO OPEN> -e source 30
    org.jellyfin.androidtv/.ui.startup.StartupActivity
target:
  entity_id: <YOUR ANDROID TV>
```

**Reply by pheffner** [reply]
Have a look at the [Android TV Remote](https://www.home-assistant.io/integrations/androidtv_remote) doc.
They show you how to call applications by their application ID. If you add a media card for the media player entity for your tv anywhere in your setup, it shows the application ID for the current app running on your TV. If you use your remote to start an app, you can find the appID that way.

The appID for the Jellyfin application is org.jellyfin.androidtv, so you can call it with yaml like this:

```
# Launch the Jellyfin app
action: media_player.play_media
data:
  media_content_type: app
  media_content_id: org.jellyfin.androidtv
target:
  entity_id: media_player.<your androidtv>
```

Works great for me, I hope this helps.

**Reply by sysmoon14** [reply]
> It depends on what you want to achieve. If you **just want to open the app** on your android tv, you could use the following adb command:
> 
> ```
> monkey -p org.jellyfin.androidtv 1
> ```
> 
> The resulting Home Assistant step would be:
> 
> ```
> service: androidtv.adb_command
> metadata: {}
> data:
>   command: monkey -p org.jellyfin.androidtv 1
> target:
>   entity_id: <YOUR ANDROID TV>
> ```
> 
> But if you, like me, would like to **open a specific title in the app**, you could use the following adb command:
> 
> ```
>  adb shell am start -c android.intent.category.LEANBACK_LAUNCHER -a android.intent.action.VIEW -d <ID OF THE TITLE YOU WANT TO OPEN> -e source 30 org.jellyfin.androidtv/.ui.startup.StartupActivity
> ```
> 
> Here the resulting Home Assistant step would be:
> 
> ```
> service: androidtv.adb_command
> data:
>   command: >-
>     adb shell am start -c android.intent.category.LEANBACK_LAUNCHER -a
>     android.intent.action.VIEW -d <ID OF THE TITLE YOU WANT TO OPEN> -e source 30
>     org.jellyfin.androidtv/.ui.startup.StartupActivity
> target:
>   entity_id: <YOUR ANDROID TV>
> ```

@ebinf This is gold! Do you know how we could find other intents? View is great, but how about Play? Any way to select a user as part of the intent? etc.

Also, do you know a way to programatically get the Jellyfin IDs from within HA?

**Reply by cad1231** [reply]
@sysmoon14 To get Jellyfin IDs programmatically with HA, look into the official HA Jellyfin addon: the [integration page](https://www.home-assistant.io/integrations/jellyfin/#action-search-media) explains exactly how to do that.

As for playing the media, I am not sure if there is a better way, but I just took the code that @ebinf provided to pull up specific media. When you do that, it defaults to having the "play" button selected. From there, it's just `adb input keyevent 66`, which programmatically presses the "enter" button at the center of any remote control D-pad -- and the selected media starts playing. 

This automatically turns on a show for my kids because they're too young to navigate 10-ft interfaces.
